src/yunohost/regenconf.py

```diff
@@ -0,0 +1,518 @@
+# -*- coding: utf-8 -*-
+
+""" License
+
+    Copyright (C) 2018 YunoHost
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as published
+    by the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program; if not, see http://www.gnu.org/licenses
+
+"""
+
+import os
+import time
+import yaml
+import json
+import subprocess
+import errno
+import shutil
+import hashlib
+
+from difflib import unified_diff
+
+from moulinette import m18n
+from moulinette.core import MoulinetteError
+from moulinette.utils import log, filesystem
+from moulinette.utils.filesystem import read_file
+
+from yunohost.hook import hook_callback
+
+BASE_CONF_PATH = '/home/yunohost.conf'
+BACKUP_CONF_DIR = os.path.join(BASE_CONF_PATH, 'backup')
+PENDING_CONF_DIR = os.path.join(BASE_CONF_PATH, 'pending')
+REGEN_CONF_FILE = '/etc/yunohost/regenconf.yml'
+
+logger = log.getActionLogger('yunohost.regenconf')
+
+
+def regen_conf(names=[], with_diff=False, force=False, dry_run=False,
+               list_pending=False):
+    """
+    Regenerate configuration file(s)
+
+    Keyword argument:
+        names -- Categories to regenerate configuration of
+        with_diff -- Show differences in case of configuration changes
+        force -- Override all manual modifications in configuration files
+        dry_run -- Show what would have been regenerated
+        list_pending -- List pending configuration files and exit
+
+    """
+
+    # Legacy code to automatically run the migration
+    # This is required because regen_conf is called before the migration call
+    # in debian's postinst script
+    if os.path.exists("/etc/yunohost/installed") \
+       and ("conffiles" in read_file("/etc/yunohost/services.yml") \
+            or not os.path.exists("/etc/yunohost/regenconf.yml")):
+        from yunohost.tools import _get_migration_by_name
+        migration = _get_migration_by_name("separate_regenconf_from_services")
+        migration.migrate()
```

Comment:
I don't know well the migration system. Is there a risk that a user running the regen-conf migrate accidentally to stretch ?

Reply:
Normally no, some migrations needs to be explicitly accepted iirc, @alexAubin will confirm.

Reply:
This is to be migrated in stretch-unstable, so implicitly people will be already on stretch